docker_swarm_facts: New module to retrieve Docker Swarm information

WojciechowskiPiotr opened the pull request:
##### SUMMARY
A new module to read facts about the swarm cluster. It is meant to replace the `state: inspect` in `docker_swarm` module for consistency across modules as discussed in #50428. 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
docker_swarm_facts

##### ADDITIONAL INFORMATION
The module returns the same data as `/swarm` API method. There is no straightforward equivalent in CLI. It returns the Swarm stack information including JoinTokens required for new nodes to join the Swarm cluster. 

Additionally, it always returns the host role or `unknown` in case host is not part of a swarm to simplify playbooks, by example looping through inventory to find managers and get JoinToken in one task.

It uses AnsibleDockerSwarmClient class which extends AnsibleDockerClient therefor this PR depends on approval of #50584 where this class is introduced in module_utils/docker_swarm.

felixfontein commented:
Since this needs #50584, you should rebase it so it includes #50584; once #50584 is merged, you can rebase against `devel` and remove the `[WIP]`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,187 @@
+node_role:
+  description:
+    - Information on role of docker host where M(docker_swarm_facts) is executed.
+    - Will be C(Manager) if executed on Swarm manager.
```
Why not use lower-case identifiers? (As you use for option values for `docker_node`, for example?)

WojciechowskiPiotr replied:
It is not returning the role anymore, looks useless when we have other modules to get that

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,187 @@
+    description:
+      - If C(short) and I(list) is C(nodes) then returns list of registered nodes names.
+      - If C(long) and I(list) is C(nodes) then returns list as in output of `docker nodes ls`.
+      - Ignored if I(list) is not present.
```
How about having only one option, `list`, with values `nodes` and `node-hostnames`?

WojciechowskiPiotr replied:
You are right, it is a better approach. But we need a better name than `node-hostnames`. It is misleading because the current `long` output returns five to seven attributes for each node depending on its role and state. Maybe `nodes-info` or `nodes-basics`?

felixfontein replied:
`node-hostnames` (for me) would be the list of hostnames, while `nodes` would be the list with more infos. Why not return the list of inspect results, though?

How about three modes:
 - `nodes`: list of inspect results
 - `node-hostnames`: list of hostnames
 - `node-summaries`: list of summaries (ID, hostname, status, ...)

WojciechowskiPiotr replied:
How about following with the approach from the `docker_host_facts`?

~~`swarm:` - swarm info~~ Swarm facts always will be part of a returned data
`nodes:` - lists nodes with basic info like in `docker node ls`
`nodes_filters:` - filtering the query range
`tasks:`- lists tasks
`tasks_filters:` - filtering the query range
`services:` - lists docker services running on swarm
`services_filters:` - filtering the query range
`verbose_output:` - instead if matching the CLI output it will return everything it gets from the API call

This way we will have a unified approach between `docker_host_facts` and `docker_swarm_facts`.

jwitko replied:
I'm a fan of this approach @WojciechowskiPiotr for this situation.

felixfontein replied:
I like that idea as well!

Commit:
docker_swarm_facts: PR cleanup after rebasing

Commit:
docker_swarm_facts: Adding the features similar to docker_host_facts

WojciechowskiPiotr commented:
ready_for_review

Please take a look, it will need small adjustments after #51700 merge but the core of the module code is not dependent on this PR.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,319 @@
+    description:
+      - When set to C(yes) and I(nodes), I(services) or I(tasks) is set to C(yes)
+        then output will contain verbose information about objects matching the full output of API method.
+        For details see the documentation of your version of Docker API at L(https://docs.docker.com/engine/api/).
```
```suggestion
        For details see the documentation of your version of Docker API at U(https://docs.docker.com/engine/api/).
```
`L(...,...)` is for links with titles. ([Details](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation))

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,319 @@
+    type: dict
+  verbose_output:
+    description:
+      - When set to C(yes) and I(nodes), I(services) or I(tasks) is set to C(yes)
```
Why not write something like `When set to C(yes), the output for I(nodes), I(services) and I(tasks) will contain verbose information ...`?

WojciechowskiPiotr replied:
This is a modified version of description in `docker_host_facts` - used the same template agreed on `docker_node_facts` module for consistency as the module usage is similar

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,319 @@
+'''
+
+RETURN = '''
+docker_swarm_facts:
```
How about removing the `docker_` prefix from the return variables, and simplifying the names in general? I would use:
- `swarm_facts` or just `swarm`
- `nodes`
- `services`
- `tasks`

WojciechowskiPiotr replied:
That will break the naming convention of returned values, I am not sure if this is a good idea. Most `docker_*_facts` modules returns variables with `docker_` prefix. Maybe we should open discussion in agenda and unify naming across modules before 2.8 freeze?

felixfontein replied:
Funny thing, the only module which was there already in Ansible 2.7 is `docker_image_facts`, and its return value is called `images` :-) Let's change it in all `_facts` modules, then. I'll create a PR for that.

WojciechowskiPiotr replied:
I am not so sure it is a good approach. I know there is no standard across modules, but it doesn't hurt to have the prefix and it may be useful if you do some manipulations from multiple `_facts` modules from different clouds.

felixfontein replied:
See #51939.

felixfontein replied:
I disagree: if you (as the module's user) want to make sure that you have something around which reminds you of docker, why not put it into the `register:` variable's name then?

felixfontein replied:
Also, let's continue the discussion in #51939.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,319 @@
+    pass
+
+try:
+    from ansible.module_utils.docker_common import docker_version, clean_dict_booleans_for_docker_api
```
There's no need to wrap this in a `try`/`except`.

WojciechowskiPiotr replied:
Will fix when after rebase due to #51700

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,319 @@
+  - Retrieves facts about a Docker Swarm.
+  - Returns lists of swarm objects names for the services - nodes, services, tasks.
+  - The output differs depending on API version available on docker host.
+  - Must be run on Swarm Manager node.
```
I think the module should also return something sensible (which is not an error) when it is not a manager node, or the docker daemon is not in swarm mode at all. Or, if that makes implementation too complicated, `docker_host_facts` should return something like that :-)

WojciechowskiPiotr replied:
Why not fail? `docker_node_facts` fails in cases it is not run on Swarm Manager. It may be easier to handle conditional in playbook if it fails

felixfontein replied:
The main reason is that usually, `xxx_facts` modules also return information on whether `xxx` is available at all. It would be fine for me to simply return more detailed error messages on failure, after all we can pass arbitrary information along when calling `module.fail_json()`. If we return this information (docker in swarm mode / node is manager) also on failure situations, users can add `ignore_errors: yes` and use that information if they like.

WojciechowskiPiotr replied:
Ok, will extend the code in such way

felixfontein replied:
Are you still working on this? I can't find anything like that in the current version.

WojciechowskiPiotr replied:
Line `168` calls the `fail_task_if_not_swarm_manager()`, I also updated the error message in the library to be more informative and consistent so ever swarm module fails the same way if not run on a manager

felixfontein replied:
Ok, but that's only the message, which shouldn't be parsed / understood by the calling code. Something useable would be a boolean variable `docker_in_swarm_mode` or `swarm_node_is_manager` which is returned in any case (except maybe when the docker daemon doesn't seem to be there in the first place).

It might be a good idea to introduce a general `fail_results` variable in `AnsibleDockerClient`, which is empty by default and passed on to `module.fail_json` in `fail()`. If everyone (i.e. also the code in module_utils/docker/swarm.py) calls `client.fail()` instead of `module.fail_json()`, it will always be returned.

Then the facts module can put the booleans into `client.fail_results` with value `False`, and update them to `True` once the corresponding calls succeed.

If you want I can prepare such a system in another PR.

WojciechowskiPiotr replied:
That is an interesting idea and can be later implemented in `fail_task_if_not_swarm_manager()` and other `AnsibleDockerSwarmClient`. In such a case, this new feature should not block this PR.

felixfontein replied:
Fine for me. 2.8 is still some some time out, fortunately :)

felixfontein replied:
See #51999 for the promised PR. (I also added some info on the 2.8 timeline to the agenda/task list; this probably interests everyone and shouldn't be buried in a comment here ;) )

WojciechowskiPiotr replied:
Good :) If nobody else has anything to add let's put ansibot to work to set required labels to merge it :)

felixfontein commented:
#51700 is merged, feel free to rebase and adjust.

Commit:
docker_swarm_facts: Minor documentation and documentation updates

Commit:
docker_swarm_facts: Minor documentation and documentation adjustments to changes in #51700

Commit:
docker_swarm_facts: Using AnsibleDockerSwarmClient class method to fail module if not run on swarm manager
module_utils\docker\swarm.py: Adjustment of error message to me bore descriptive if module fails because it is not run on swarm manager node

WojciechowskiPiotr commented:
ready_for_review

felixfontein commented:
Looks good. Same as for #51701: if no-one complains, I'll merge by Tuesday.

felixfontein commented:
shipit

felixfontein commented:
@WojciechowskiPiotr and also thanks for this one!
@jwitko thanks for your feedback!

WojciechowskiPiotr commented:
> @WojciechowskiPiotr and also thanks for this one!
> @jwitko thanks for your feedback!

My pleasure
